src/hotspot/share/gc/g1/g1Analytics.cpp

```diff
@@ -317,6 +326,18 @@ double G1Analytics::most_recent_gc_end_time_sec() const {
 
 void G1Analytics::update_recent_gc_times(double end_time_sec,
                                          double pause_time_ms) {
+  // If a Remark and/or Cleanup pause happened leading up to this pause,
+  // piggy-back the Remark or Cleanup pause time onto this pause.
+  // Reason why we do not update_recent_gc_times at the end of Remark/Cleanup is
+  // that allocation rates are computed using recent_gc_times as checkpoints.
+  // Updating on Remark/Cleanup would introduce check points that didn't do any
+  // collections.
+  double most_recent_gc_end_ms =  most_recent_gc_end_time_sec() * 1000.0;
+  assert(_prev_collection_pause_end_ms >= most_recent_gc_end_time_ms, "should be");
+  double remark_cleanup_pause_ms = _prev_collection_pause_end_ms - most_recent_gc_end_ms;
+
+  pause_time_ms += remark_cleanup_pause_ms;
+
```

Comment:
Feels like we're doing the same thing over again. Could we somehow restructure/re-design this to avoid doing the same thing twice. Maybe we should look at recording the remark/cleanup pause-time instead of appending them to the end of a normal gc.

Reply:
I agree, a complete restructure/re-design would be better. However, recording the remark/cleanup pause-times in the current set up would break some assumptions: "Reason why we do not update_recent_gc_times at the end of Remark/Cleanup is that allocation rates are computed using recent_gc_times as checkpoints. With the assumption that collections have happened at these checkpoints. Updating on Remark/Cleanup would introduce check points that didn't do any collections"

Reply:
I was a bit unclear, I should not have used "record". We could still add a separate member for this in G1Analytics, but still not add the checkpoints.

Reply:
ok